src/Cli/dotnet/commands/dotnet-workload/install/NetSdkManagedInstaller.cs

```diff
@@ -0,0 +1,283 @@
+
+        public override void DownloadToOfflineCache(IReadOnlyCollection<string> manifests) => throw new NotImplementedException();
+
+        public override void GarbageCollectInstalledWorkloadPacks()
+        {
+            var sdkFeatureBands = GetFeatureBandsWithInstallationRecords();
+            _reporter.WriteLine(string.Format(LocalizableStrings.GarbageCollectingSdkFeatureBandsMessage, string.Join(", ", sdkFeatureBands)));
+            var deletablePacks = GetDeletablePacks(sdkFeatureBands);
+
+            foreach (var (deletablePack, featureBand) in deletablePacks)
+            {
+                DeletePackInstallationRecord(deletablePack, featureBand);
+                DeletePack(deletablePack);
+            }
+        }
+
+        private IEnumerable<(PackInfo, string)> GetDeletablePacks(IEnumerable<string> sdkFeatureBands)
+        {
+            var installedPacksDir = Path.Combine(_workloadMetadataDir, _installedPacksDir);
+            IEnumerable<(PackInfo, string)> deletablePacks = new List<(PackInfo, string)>();
+            foreach (var featureBand in sdkFeatureBands)
+            {
+                var workloadResolver = GetWorkloadResolver(featureBand);
+                var installedWorkloads = GetInstalledWorkloads(featureBand);
+
+                var expectedPacks = installedWorkloads
+                    .SelectMany(workload => workloadResolver.GetPacksInWorkload(workload))
+                    .Select(pack => workloadResolver.TryGetPackInfo(pack))
+                    .Select(packInfo => GetPackInstallRecordPath(packInfo, featureBand));
+
+                var currentFeatureBandPackRecords = Directory.EnumerateDirectories(installedPacksDir)
+                    .SelectMany(packIdDir => Directory.EnumerateDirectories(packIdDir))
+                    .SelectMany(packVersionDir => Directory.EnumerateFiles(packVersionDir));
+
+                currentFeatureBandPackRecords = currentFeatureBandPackRecords
+                    .Where(featureBandDir => Path.GetFileName(featureBandDir).Equals(featureBand));
+
+                var featureBandPacksToDelete = currentFeatureBandPackRecords.Except(expectedPacks)
+                    .Select(packRecordPath => (GetPackInfo(workloadResolver, packRecordPath), featureBand));
+
+                deletablePacks = deletablePacks.Concat(featureBandPacksToDelete);
+            }
+            return deletablePacks;
+        }
+
+        private PackInfo GetPackInfo(WorkloadResolver workloadResolver, string packRecordPath)
+        {
+            var versionRecordPath = Path.GetDirectoryName(packRecordPath);
+            var idRecordPath = Path.GetDirectoryName(versionRecordPath);
+            var packId = Path.GetFileName(idRecordPath);
+            return workloadResolver.TryGetPackInfo(packId);
+        }
+
+        protected virtual WorkloadResolver GetWorkloadResolver(string featureBand)
```

Comment:
constructor injection is not enough?

usually we do this for legacy code when the separation of functionality is not intentional

Reply:
Yeah, that's something I wanted to check, probably with @dsplaisted. The workload resolvers are by feature band (in that they load and work off of only the manifests in the feature band we specify), right? If so, then we'll have to create one for each feature band when we garbage collect here: https://github.com/dotnet/sdk/pull/16530/files#diff-c01acc54f7e26514dc3843cef3eb6b1418943a3e139ea89abbaa27af964dc39bR128